pkg/kubelet/status/status_manager.go

```diff
@@ -771,25 +832,47 @@ func normalizeStatus(pod *v1.Pod, status *v1.PodStatus) *v1.PodStatus {
 	return status
 }
 
-// mergePodStatus merges oldPodStatus and newPodStatus where pod conditions
-// not owned by kubelet is preserved from oldPodStatus
-func mergePodStatus(oldPodStatus, newPodStatus v1.PodStatus) v1.PodStatus {
-	podConditions := []v1.PodCondition{}
+// mergePodStatus merges oldPodStatus and newPodStatus to preserve where pod conditions
+// not owned by kubelet and to ensure terminal phase transition only happens after all
+// running containers have terminated. This method does not modify the old status.
+func mergePodStatus(oldPodStatus, newPodStatus v1.PodStatus, couldHaveRunningContainers bool) v1.PodStatus {
+	podConditions := make([]v1.PodCondition, 0, len(oldPodStatus.Conditions)+len(newPodStatus.Conditions))
+
 	for _, c := range oldPodStatus.Conditions {
 		if !kubetypes.PodConditionByKubelet(c.Type) {
 			podConditions = append(podConditions, c)
 		}
 	}
-
 	for _, c := range newPodStatus.Conditions {
 		if kubetypes.PodConditionByKubelet(c.Type) {
 			podConditions = append(podConditions, c)
 		}
 	}
 	newPodStatus.Conditions = podConditions
+
+	// Delay transitioning a pod to a terminal status unless the pod is actually terminal.
+	// The Kubelet should never transition a pod to terminal status that could have running
+	// containers and thus actively be leveraging exclusive resources. Note that resources
+	// like volumes are reconciled by a subsystem in the Kubelet and will converge if a new
+	// pod reuses an exclusive resource (unmount -> free -> mount), which means we do not
+	// need wait for those resources to be detached by the Kubelet. In general, resources
+	// the Kubelet exclusively owns must be released prior to a pod being reported terminal,
+	// while resources that have participanting components above the API use the pod's
+	// transition to a terminal phase (or full deletion) to release those resources.
+	if !isPhaseTerminal(oldPodStatus.Phase) && isPhaseTerminal(newPodStatus.Phase) {
+		if couldHaveRunningContainers {
+			newPodStatus.Phase = oldPodStatus.Phase
```

Comment:
I think we may have to should preserve the old message and reason as well:

```
newPodStatus.Message = oldPodStatus.Message
newPodStatus.Reason = oldPodStatus.Reason
```

I did some testing of this PR with conjunction of graceful shutdown and the noticed the following behavior:

During shutdown, I noticed that the first status update would be sent which updated the [message and reason](https://github.com/kubernetes/kubernetes/blob/9e5efeb/pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go#L323-L324) for the pod. The message was set to `Terminated`, which makes it look like the pod is already terminated, but the phase is still running. Only after all the containers terminate (which is terminationGracePeriod seconds later), will the second status update be sent with `Phase=Failed`. 

This is a bit confusing, since the pod can be stuck with `Terminated` message, but still be running if the node happens to shutdown before the second status update is sent. This is also behavior change, since previously phase=Failed and message/reason updates were sent in a single update request.

To fix it, perhaps we should also apply the same logic of using the oldPhase to the message and reason.

Reply:
Interesting.  So this is two factors competing:

1. As a user, I want to know that kubelet is responding (post this change, it takes longer to know that)
2. As a user, I want to be able to see consistent messaging (post this change, message/reason and phase were not aligned)

The best option (in a follow up) is to have a couple of new conditions - `Terminating` and `Evicted` that capture one way transitions (once the condition is present, there is only one outcome) and cleanly solve 1.  Since that's a longer change, I agree that message and reason should probably be aligned in this pr.  I could foresee some user regression in that (especially for eviction) but that's just a reason to prioritize the conditions.

Reply:
Makes sense. I agree long term we probably need an intermediate state (i.e. `Terminating...`), but for now I think it'll best to align message and reason in a single update.